### DISCUSSION TITLE
Query about Fastest and Synchronized way of MPU6050s reading by multiplexer

### DISCUSSION THREAD

**Discussion by williamesp2015** [jrowberg/i2cdevlib#628] [General]
I have problem with reading fast and efficient way to read multiple MPU6050 sensors using Multiplexer. Finally I found the fastest way is reading RAW data but is not synchronized because reading each sensor taking 1-2 msec but if I could start all the sensor FIFO by same sampling rate then can read sequentially one by one.

**Comment by ZHomeSlice** [comment]
how to start DMP FIFO at the same time is discussed here:
https://github.com/jrowberg/i2cdevlib/discussions/621#discussioncomment-1488519

Z

**Comment by williamesp2015** [comment]
MPU6050_6Axis_MotionApps612.cpp is easy to divide into all sensor initialization and second part for each sensor but due to my difficulty(DeviceID=0xC) in the 612 version I have to use MPU6050_6Axis_MotionApps20.cpp and it is not suitable for this purpose.

**Reply by ZHomeSlice** [reply]
V2.0 is still a mystery to me on how to configure it. Noah Zerkin who helped develop the MotionApps612.cpp is mentioned but I believe he is not available to help resolve your need.

 V2.0 note in MPU6050_6Axis_MotionApps20.cpp
```
 * Also, I'd like to offer many, many thanks to Noah Zerkin for all of the
 * DMP reverse-engineering he did to help make this bit of wizardry
 * possible.
```

question:
(DeviceID=0xC)
is it different when you use V6.12 vs V2.0?

Z

**Comment by williamesp2015** [comment]
DeviceID=0xC simmilar in V2-0 and V6-12. In V6.12 with modifies getDeviceID() that always return 0x34, ESP32 always stop in the beginning of loop() mpu.dmpGetCurrentFIFOPacket(fifoBuffer).

**Reply by ZHomeSlice** [reply]
@williamesp2015 
You are saying that the ESP32 freezes when using mpu.dmpGetCurrentFIFOPacket(fifoBuffer).
??

possible solution:
Have you tried adding a simple delay(1); before and or after dmpGetCurrentFIFOPacket to allow the esp32 to do wifi stuff in the background?

Z

**Reply by williamesp2015** [reply]
Yes, I put delay in the loop:
while(!dmpGetCurrentFIFOPacket )
{
delay(1);
}

**Reply by ZHomeSlice** [reply]
@williamesp2015 
V2.0 is so deprecated that I have not modified or updated the code for retrieving the data from the FIFO buffer in the examples

this function should be available to you:
```
/** Get the latest byte from the FIFO buffer no matter how much time has passed.
 * ===                  GetCurrentFIFOPacket                    ===
 * ================================================================
 * Returns 1) when nothing special was done
 *         2) when recovering from the overflow
 *         0) when no valid data is available
 * ================================================================ */
 int8_t MPU6050::GetCurrentFIFOPacket(uint8_t *data, uint8_t length) { // overflow proof
     int16_t fifoC;
     // This section of code is for when we allowed more than 1 packet to be acquired
     //uint32_t BreakTimer = micros();
     uint32_t BreakTimer = time_us_32();
     do {
         if ((fifoC = getFIFOCount())  > length) {

             if (fifoC > 200) { // if you waited to get the FIFO buffer to > 200 bytes it will take longer to get the last packet in the FIFO Buffer than it will take to  reset the buffer and wait for the next to arrive
                 resetFIFO(); // Fixes any overflow corruption
                 fifoC = 0;
                 while (!(fifoC = getFIFOCount()) && ((time_us_32() - BreakTimer) <= (11000))); // Get Next New Packet
                 } else { //We have more than 1 packet but less than 200 bytes of data in the FIFO Buffer
                 uint8_t Trash[BUFFER_LENGTH];
                 while ((fifoC = getFIFOCount()) > length) {  // Test each time just in case the MPU is writing to the FIFO Buffer
                     fifoC = fifoC - length; // Save the last packet
                     uint16_t  RemoveBytes;
                     while (fifoC) { // fifo count will reach zero so this is safe
                         RemoveBytes = MIN((int)fifoC, BUFFER_LENGTH); // Buffer Length is different than the packet length this will efficiently clear the buffer
                         getFIFOBytes(Trash, (uint8_t)RemoveBytes);
                         fifoC -= RemoveBytes;
                     }
                 }
             }
         }
         if (!fifoC) return 0; // Called too early no data or we timed out after FIFO Reset
         // We have 1 packet
         if ((time_us_32() - BreakTimer) > (11000)) return 0;
     } while (fifoC != length);
     getFIFOBytes(data, length); //Get 1 packet
     return 1;
}
```
This code will pick approximately the best way to get the latest packet. eliminating the timeout issues you face. 
Z
